snooty/semanticparser.py

```diff
@@ -85,201 +77,167 @@ def build_slug_title_mapping(
             assert children is not None
             self.slug_title_mapping[slug] = children
 
-    def build_toctree(self, pages: Dict[FileId, Page]) -> Dict[str, SerializableType]:
-        fileid_dict = {}
-        toctree: Dict[str, Any] = {"toctree": {}}
+    def build_slug_fileid_mapping(self) -> None:
+        """Construct a {slug: fileid} mapping so that we can retrieve the full file name
+        given a slug. We cannot use the with_suffix method since the type of the slug
+        in find_toctree_nodes(...) is string rather than FileId."""
+        fileid_dict: Dict[str, FileId] = {}
+        for fileid in self.pages:
+            slug = fileid.without_known_suffix
+            fileid_dict[slug] = fileid
+        self.slug_fileid_mapping = fileid_dict
+
+    def build_toctree(self) -> Dict[str, SerializableType]:
+        """Build property toctree"""
 
         # The toctree must begin at either `contents.txt` or `index.txt`.
         # Generally, repositories will have one or the other; but, if a repo has both,
         # the starting point will be `contents.txt`.
         candidates = (FileId("contents.txt"), FileId("index.txt"))
         starting_fileid = next(
-            (candidate for candidate in candidates if candidate in pages), None
+            (candidate for candidate in candidates if candidate in self.pages), None
         )
         if starting_fileid is None:
             return {}
 
-        # Construct a {slug: fileid} mapping so that we can retrieve the full file name
-        # given a slug. We cannot use the with_suffix method since the type of the slug
-        # in find_toctree_nodes(...) is string rather than FileId.
-        for fileid in pages:
-            slug = fileid.without_known_suffix
-            fileid_dict[slug] = fileid
-
         if not self.slug_title_mapping:
-            self.run_event_parser(pages)
+            self.run_event_parser()
```

Comment:
I see a couple of these `if not self.foo: <generate-self.foo>` constructs. Is this class intended to be used in a one-shot sort of way? If so, we probably want to mention that in the class docstring.